src/sumcheck/sumcheck_polynomial.rs

```diff
@@ -119,13 +126,88 @@ where
     /// incorrect results for polynomials generated by the univariate skip optimization.
     #[must_use]
     #[inline]
-    pub fn evaluate_on_standard_domain(&self, point: &MultilinearPoint<F>) -> F {
+    pub fn evaluate_on_standard_domain(&self, point: &MultilinearPoint<F>) -> F
+    where
+        F: Send + Sync,
+    {
         assert_eq!(point.num_variables(), self.n_variables);
-        self.evaluations
-            .iter()
-            .enumerate()
-            .map(|(i, &eval)| eval * point.eq_poly3(i))
-            .sum()
+        let n_vars = self.n_variables;
+        if n_vars == 0 {
+            // If there are no variables, the polynomial is a constant.
+            return self.evaluations.first().copied().unwrap_or(F::ZERO);
+        }
+
+        // Ping-Pong Buffer Strategy:
+        //
+        // We allocate two buffers upfront to avoid allocations in the loop.
+
+        // Start with the full set of evaluations.
+        let mut current_evals = self.evaluations.clone();
+        // The destination buffer for the next folded state.
+        let mut next_evals = vec![F::ZERO; self.evaluations.len() / 3];
+
+        // Iteratively fold the evaluation table one dimension at a time.
+        //
+        // We iterate in reverse to process from the last variable (fastest-moving index) to the first.
+        for (i, &p_i) in point.iter().rev().enumerate() {
+            let p_i_min_1 = p_i - F::ONE;
+            let p_i_sq_min_p_i = p_i * p_i_min_1;
+            let lagrange_evals = [
+                // L_0(p_i) = (p_i - 1)(p_i - 2) / 2
+                //
+                // (p_i - 1)(p_i - 2) / 2
+                // = (p_i^2 - p_i + 2)/2
+                // = (p_i^2 - p_i - p_i + 2)/2
+                // = (p_i(p_i-1) - 2(p_i-1))/2
+                // = p_i(p_i-1)/2 - (p_i-1)
+                p_i_sq_min_p_i.halve() - p_i_min_1,
+                // L_1(p_i) = -p_i(p_i - 2)
+                //
+                // -p_i * (p_i - 2)
+                // = -p_i^2 + p_i
+                // = -p_i(p_i-1) + val
+                p_i - p_i_sq_min_p_i,
+                // L_2(p_i) = p_i(p_i - 1) / 2
+                //
+                // p_i * (p_i - 1) / 2
+                p_i_sq_min_p_i.halve(),
+            ];
+
+            // Determine the source and destination buffers for this iteration.
+            let (source, dest) = if i % 2 == 0 {
+                (&current_evals, &mut next_evals)
+            } else {
+                (&next_evals, &mut current_evals)
+            };
+
+            // Ensure the destination buffer is the correct size for this folding step.
+            let next_size = source.len() / 3;
+            dest.resize(next_size, F::ZERO);
```

Comment:
Hmm, this should always shrink `dest` right?
```suggestion
            dest.truncate(next_size);
```